### PR TITLE
addresses #495 and then some

### PR DESCRIPTION
**Pull request by alexlamsl**
In the process of refactoring for #495, I've discovered a bug when `conservativeCollapse` and `preserveLineBreaks` are used on elements with only `\n` on one side but whitespaces on the other.

### CODE REVIEW

**Comment by alexlamsl** [comment]
I think by integrating the default value into `ignoreCustomFragments` in the docs, we also reduce the number of instances where users didn't realise `<?php ?>` is already accounted for, like #489 and others before it...

**Comment by alexlamsl** [comment]
Now that I've worked with `ignoreCustomFragments` for a while, it's starting to look like a golden hammer :laughing: 

On a more serious note, contemplating on whether to use this for IE Conditional Comments...

**Comment by alexlamsl** [comment]
Checked with `node benchmark` and no noticable regression in sizes or performance has been detected.

**Comment by alexlamsl** [comment]
Travis CI tripped :confused:

File: dist/htmlminifier.js [hunk 1]
```diff
@@ -618,13 +593,10 @@
     return createMap(values.split(/,/));
   }
 
-  // array of non-empty element tags that will maintain a single space outside of them
-  var inlineTags = createMapFromString('a,abbr,acronym,b,bdi,bdo,big,button,cite,code,del,dfn,em,font,i,ins,kbd,mark,math,q,rt,rp,s,samp,small,span,strike,strong,sub,sup,svg,time,tt,u,var');
-
-  function collapseWhitespaceSmart(str, prevTag, nextTag, options) {
+  function collapseWhitespace(str, preserveLineBreaks, conservativeCollapse, trimLeft, trimRight, collapseAll) {
```
Comment (kangax) [inline dist/htmlminifier.js]: 6 args is kinda scary; can we make it prettier? :)

File: dist/htmlminifier.js [hunk 2]
```diff
@@ -634,24 +606,32 @@
...
 
     return lineBreakBefore + str + lineBreakAfter;
   }
 
+  // array of non-empty element tags that will maintain a single space outside of them
+  var inlineTags = createMapFromString('a,abbr,acronym,b,bdi,bdo,big,button,cite,code,del,dfn,em,font,i,ins,kbd,mark,math,q,rt,rp,s,samp,small,span,strike,strong,sub,sup,svg,time,tt,u,var');
+  var selfClosingInlineTags = createMapFromString('comment,img,input');
+
+  function collapseWhitespaceSmart(str, prevTag, nextTag, options) {
+    var trimLeft = prevTag && !selfClosingInlineTags(prevTag) && (options.collapseInlineTagWhitespace || prevTag.charAt(0) !== '/' || !inlineTags(prevTag.substr(1)));
```
Comment (kangax) [inline dist/htmlminifier.js]: Just need a little better formatting here (for easier reading)

File: dist/htmlminifier.js [hunk 3]
```diff
@@ -1508,8 +1531,8 @@
           }
           optionalStartTag = '';
           // </html> or </body> may be omitted if not followed by comment
-          // </head> or </colgroup> may be omitted if not followed by space or comment
-          if (optionalEndTag === 'html' || optionalEndTag === 'body' || (optionalEndTag === 'head' || optionalEndTag === 'colgroup') && !/^\s/.test(text)) {
+          // </head>, </colgroup> or </caption> may be omitted if not followed by space or comment
+          if (optionalEndTag === 'html' || optionalEndTag === 'body' || (optionalEndTag === 'head' || optionalEndTag === 'colgroup' || optionalEndTag === 'caption') && !/^\s/.test(text)) {
```
Comment (kangax) [inline dist/htmlminifier.js]: Is scoping with parens necessary here?

Comment (alexlamsl) [inline dist/htmlminifier.js]: If you mean the parens around `head`, `colgroup` and `caption`, then yes, due to precedence of `&&` over `||`, i.e. `!/^\s/.test(text)` applies to all of those inside the parens, but not `head` or `body`.

But I guess another `createMapFromString` is in store here... :stuck_out_tongue_winking_eye:

**Comment by alexlamsl** [comment]
@kangax PTAL

File: dist/htmlminifier.js [hunk 4]
```diff
@@ -1568,8 +1595,12 @@
...
 
     if (uidAttr) {
-      str = str.replace(new RegExp('\\s*' + uidAttr + '\\s*', 'g'), function() {
-        return ignoredCustomMarkupChunks.shift();
+      str = str.replace(new RegExp('(\\s*)' + uidAttr + '(\\s*)', 'g'), function(match, prefix, suffix) {
+        var chunk = ignoredCustomMarkupChunks.shift();
+        return options.collapseWhitespace ? collapseWhitespace(prefix + chunk + suffix, {
+          preserveLineBreaks: options.preserveLineBreaks,
+          conservativeCollapse: true
+        }, true, true) : chunk;
```
Comment (alexlamsl) [inline dist/htmlminifier.js:1603]: OT: I wonder how far modern JavaScript engines can optimise this to:

```
if (options.collapseWhitespace) {
  str = str.replace(new RegExp('\\s*' + uidAttr + '\\s*', 'g'), function() {
    return ignoredCustomMarkupChunks.shift();
  });
}
else {
  str = str.replace(new RegExp('(\\s*)' + uidAttr + '(\\s*)', 'g'), function(match, prefix, suffix) {
    var chunk = ignoredCustomMarkupChunks.shift();
    return collapseWhitespace(prefix + chunk + suffix, {
      preserveLineBreaks: options.preserveLineBreaks,
      conservativeCollapse: true
    }, true, true);
  });
}
```

**Comment by kangax** [comment]
Thanks! Just to confirm, this has no backwards-incompatible changes, right?

**Comment by alexlamsl** [comment]
@kangax not intentionally, no :sweat_smile: 

For instance, the old rules for `<rtc>` etc. are kept to ensure compatibility between W3C and WhatWG versions.

**Comment by alexlamsl** [comment]
@kangax on a related note, do we have plans for publishing `1.2.0` to `npm`?

**Comment by kangax** [comment]
@alexlamsl done!

**Comment by alexlamsl** [comment]
@kangax thanks :+1:

**Pull request merged**